pkg/newstorage/newstorage.go

```diff
@@ -0,0 +1,122 @@
+
+// WithPageSize sets the maximum page size for data retrievals done within the Iterator returned by the Query call.
+// Paging is handled internally by the Iterator. Higher values may reduce CPU time and the number of database calls at
+// the expense of higher memory usage.
+func WithPageSize(size int) QueryOption {
+	return func(opts *QueryOptions) {
+		opts.pageSize = size
+	}
+}
+
+// Tag represents a Name + Value pair that can be associated with a k,v pair for querying later.
+type Tag struct {
+	Name  string
+	Value string
+}
+
+// Operation represents an operation to be performed in the Batch method.
+type Operation struct {
+	Key   string
+	Value []byte // A nil value will result in a delete operation
+	Tags  []Tag  // Optional
+}
+
+// Provider represents a storage provider.
+type Provider interface {
+	// OpenStore opens a store with the given name and returns a handle.
+	// If the store has never been opened before, then it is created.
+	OpenStore(name string) (Store, error)
+
+	// SetStoreConfig sets the configuration on a store.
+	// The store must be created prior to calling this method.
+	SetStoreConfig(name string, config StoreConfiguration) error
+
+	// GetStoreConfig gets the current store configuration.
+	// The store must be created prior to calling this method.
+	GetStoreConfig(name string) (StoreConfiguration, error)
+
+	// CloseStore closes the store with the given name.
+	// This does not delete any data in the store, and the store can be reopened again.
+	CloseStore(name string) error
+
+	// Close closes all stores created under this store provider.
+	// This does not delete any data in the stores, and the stores can be reopened again.
+	Close() error
+}
+
+// Store represents a storage database.
+type Store interface {
+	// Put stores the key + value pair along with the (optional) tags.
+	Put(key string, value []byte, tags ...Tag) error
+
+	// Get fetches the value associated with the given key.
+	Get(key string) ([]byte, error)
+
+	// Get fetches all tags associated with the given key.
+	GetTags(k string) ([]Tag, error)
+
+	// GetBulk fetches the values associated with the given keys.
+	// If a key doesn't exist, then a nil []byte is returned for that value and is not considered an error.
+	// Depending on the implementation, this method may be faster than calling Get for each key individually.
+	GetBulk(keys ...string) ([][]byte, error)
+
+	// Query returns all data that satisfies the expression. Expression format: TagName:TagValue.
+	// If TagValue is not provided, then all data associated with the TagName will be returned.
+	// For now, expression can only be a single tag Name + Value pair.
+	// If no options are provided, then defaults will be used.
+	Query(expression string, options ...QueryOption) (Iterator, error)
+
+	// Delete deletes the key + value pair (and all tags) associated with key.
+	Delete(key string) error
+
+	// Batch performs multiple Put and/or Delete operations in order.
```

Comment:
Should comment on what happens in case of operation failure(s).

How does the function indicate which operations failed (and why)?

Reply:
See https://github.com/hyperledger/aries-framework-go/pull/2426/files#diff-98db716f6d6bedd511a5f903dc6d785b6f60df130787f50612d64804a1f8195fR13 in my new PR